halo2_proofs/src/dev/failure.rs

```diff
@@ -611,6 +644,171 @@ fn render_lookup<F: Field>(
     }
 }
 
+fn render_shuffle<F: Field>(
+    prover: &MockProver<F>,
+    name: &str,
+    shuffle_index: usize,
+    location: &FailureLocation,
+) {
+    let n = prover.n as i32;
+    let cs = &prover.cs;
+    let shuffle = &cs.shuffles[shuffle_index];
+
+    // Get the absolute row on which the lookup's inputs are being queried, so we can
+    // fetch the input values.
+    let row = match location {
+        FailureLocation::InRegion { region, offset } => {
+            prover.regions[region.index].rows.unwrap().0 + offset
+        }
+        FailureLocation::OutsideRegion { row } => *row,
+    } as i32;
+
+    let shuffle_columns = shuffle.shuffle_expressions.iter().map(|expr| {
```

Comment:
Nitpick: The closure passed here is also the same as the one used in render_lookup, so it could be deduplicated.  After a careful look it seems `render_shuffle` and `render_lookup` are very similar, so I think there's an opportunity to deduplicate a lot of code.